src/ucp/proto/proto_am.inl

```diff
@@ -125,22 +108,68 @@ ucp_do_am_bcopy_single(uct_pending_req_t *self, uint8_t am_id,
 }
 
 static UCS_F_ALWAYS_INLINE ucs_status_t
-ucp_am_handle_user_header_send_status(ucp_request_t *req,
-                                      ucs_status_t send_status)
+ucp_proto_am_req_copy_header(ucp_request_t *req)
 {
-    ucs_status_t status;
+    void *user_header;
+
+    if ((req->flags & UCP_REQUEST_FLAG_USER_HEADER_COPIED) ||
+        (req->send.msg_proto.am.header.length == 0)) {
+        return UCS_OK;
+    }
 
-    if (ucs_unlikely(send_status == UCS_ERR_NO_RESOURCE) &&
+    ucs_assert(req->send.msg_proto.am.flags & UCP_AM_SEND_FLAG_COPY_HEADER);
+    user_header = ucs_mpool_set_get_inline(&req->send.ep->worker->am_mps,
+                                           req->send.msg_proto.am.header.length);
+    if (ucs_unlikely(user_header == NULL)) {
+        ucs_error("failed to allocate active message user header copy");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    memcpy(user_header, req->send.msg_proto.am.header.ptr,
+           req->send.msg_proto.am.header.length);
+    req->flags                       |= UCP_REQUEST_FLAG_USER_HEADER_COPIED;
+    req->send.msg_proto.am.header.ptr = user_header;
+
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_am_handle_user_header_send_status_common(ucp_request_t *req,
+                                             ucs_status_t status, int abort,
+                                             int release)
```

Comment:
https://github.com/openucx/ucx/pull/10452#discussion_r1939339129 refer to initial conversation